test/interpolating/partially-variable-paths.test.js

```diff
@@ -0,0 +1,123 @@
+  t.plan(16)
+
+  testCases.nonEmpty.forEach(function (testCase) {
+    t.equal(rhumb.interpolate(testCase.input, { foo: 'bing', bar: 'zap' }), testCase.expected
+      , 'returns path where partial variable substituted with their value')
+  })
+})
+
+test('Interpolating should replace empty partial variables with an empty string', function (t) {
+  t.plan(16)
+
+  testCases.empty.forEach(function (testCase) {
+    t.equal(rhumb.interpolate(testCase.input, { foo: '', bar: 'zap' }), testCase.expected
+      , 'returns path where empty values ared substituted with empty string')
+  })
+})
+
+test('Interpolating should replace null valued partial variables with an empty string', function (t) {
+  t.plan(16)
+
+  testCases.nullValued.forEach(function (testCase) {
+    t.equal(rhumb.interpolate(testCase.input, { foo: null, bar: 'zap' }), testCase.expected
+      , 'returns path where empty values ared substituted with empty string')
+  })
+})
+
+test('Interpolating should replace undefined valued partial variables with an empty string', function (t) {
+  t.plan(16)
+
+  testCases.undefinedValued.forEach(function (testCase) {
+    t.equal(rhumb.interpolate(testCase.input, { foo: undefined, bar: 'zap' }), testCase.expected
+      , 'returns path where empty values ared substituted with empty string')
+  })
+})
+
+test('Interpolating should replace missing partial variables with an empty string', function (t) {
+  t.plan(16)
+
+  testCases.missing.forEach(function (testCase) {
+    t.equal(rhumb.interpolate(testCase.input, { bar: 'zap' }), testCase.expected
+      , 'returns path where empty values ared substituted with empty string')
+  })
+})
+
+test('Interpolating should replace the values of partial variables in optional paths', function (t) {
+  t.plan(10)
+
+  t.equal(rhumb.interpolate('/wibble(/{foo}-{bar})', { foo: 'bing', bar: 'wobble' }), '/wibble/bing-wobble'
+    , 'returns path where partial variable substituted with their value')
+
+  t.equal(rhumb.interpolate('/wibble(/{foo}-part/part-{bar})', { foo: 'bing', bar: 'wobble' }), '/wibble/bing-part/part-wobble'
+    , 'returns path where partial variable substituted with their value')
+
+  ;[{ foo: null, bar: 'wobble' }, { foo: '', bar: 'wobble' }].forEach(function (param) {
+    t.equal(rhumb.interpolate('/wibble(/{foo}-{bar})', param), '/wibble/-wobble'
+      , 'returns path where empty values ared substituted with empty string')
+
+    t.equal(rhumb.interpolate('/wibble(/{foo}-part/part-{bar})', param), '/wibble/-part/part-wobble'
+      , 'returns path where empty values ared substituted with empty string')
+  })
+
+  ;[{ foo: undefined, bar: 'wobble' }, { bar: 'wobble' }].forEach(function (param) {
+    t.equal(rhumb.interpolate('/wibble(/{foo}-{bar})', param), '/wibble'
+      , 'returns path where optional parts are excluded')
+
+    t.equal(rhumb.interpolate('/wibble(/{foo}-part/part-{bar})', param), '/wibble'
+      , 'returns path where optional parts are excluded')
```

Comment:
In https://github.com/websdk/rhumb/pull/18#issuecomment-283142883 we see how RFC 6570 URI Template deals with a "similar" issue.  It seems to suggestion terms of variable and partial substitution, both empty and undefined should be a replacement with an empty string.  When it comes to generating components (paths, query params, lists, ...) then empty replacement with an empty string and undefined should be a removed / absent component.

@mstade This seems to match was we had discussed before.

--------

💭 Now do variables inside an optional part, give you the ability to create components, or is it just substitution? I think the answer is no, we should just see them as substitutions.

* I like the approach of RFC 6570 URI for generating variable length paths, however I think rhumb is currently too different for all the principles in the RTC spec to fit.
* rhumb does not provide matching like ES6 "rest" syntax.
   * 😎 Wouldn't it be cool to have the route `settings/account/{rest...}` to allow me to have a route where all decendents would be allowed to match, not just the children.
   * Now if we supported this, then I think empty and undefined would produce different urls.